File: app/src/main/java/agersant/polaris/features/browse/BrowseAdapterAlbum.java
package agersant.polaris.features.browse;

import android.util.SparseIntArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import agersant.polaris.CollectionItem;
import agersant.polaris.PlaybackQueue;
import agersant.polaris.R;
import agersant.polaris.Song;
import agersant.polaris.api.API;

import static agersant.polaris.features.browse.BrowseAdapterAlbum.AlbumViewType.DISC_HEADER;
import static agersant.polaris.features.browse.BrowseAdapterAlbum.AlbumViewType.TRACK;


class BrowseAdapterAlbum extends BrowseAdapter {

    private SparseIntArray discSizes; // Key is disc number, value is number of tracks
    private int numDiscHeaders;
    private final API api;
    private final PlaybackQueue playbackQueue;

    @Override
    void setItems(List<? extends CollectionItem> items) {
        var songs = (List<Song>) items;
        discSizes = new SparseIntArray();
        for (Song song : songs) {
            int discNumber = song.getDiscNumber();
            discSizes.put(discNumber, 1 + discSizes.get(discNumber, 0));
        }
        numDiscHeaders = discSizes.size();
        if (numDiscHeaders == 1) {
            numDiscHeaders = 0;
        }
        super.setItems(items);
    }

    BrowseAdapterAlbum(API api, PlaybackQueue playbackQueue) {
        super();
        this.api = api;
        this.playbackQueue = playbackQueue;
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + numDiscHeaders;
    }

    @Override
    public int getItemViewType(int position) {
        int currentDiscStart = 0;
        for (int discIndex = 0; discIndex < numDiscHeaders; discIndex++) {
            int currentDiscSize = discSizes.valueAt(discIndex);
            if (position == currentDiscStart) {
                return DISC_HEADER.ordinal();
            } else if (position < currentDiscStart + currentDiscSize + 1) {
                return TRACK.ordinal();
            }
            currentDiscStart += currentDiscSize + 1;
        }

        return TRACK.ordinal();
    }

    @Override
    public BrowseItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemQueueStatusView = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_browse_item_queued, parent, false);
        if (viewType == DISC_HEADER.ordinal()) {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_browse_album_disc_header, parent, false);
            return new BrowseItemHolderAlbumDiscHeader(api, playbackQueue, this, itemView, itemQueueStatusView);
        } else {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_browse_album_item, parent, false);
            return new BrowseItemHolderAlbumTrack(api, playbackQueue, this, itemView, itemQueueStatusView);
        }
    }

    @Override
    public void onBindViewHolder(BrowseItemHolder holder, int position) {

        if (holder instanceof BrowseItemHolderAlbumTrack) {

            // Assign track item
            if (numDiscHeaders > 0) {
                int offset = 1;
                int currentDiscStart = 0;
                for (int discIndex = 0; discIndex < numDiscHeaders; discIndex++) {
                    int currentDiscSize = discSizes.valueAt(discIndex);
                    if (position < currentDiscStart + currentDiscSize + 1) {
                        break;
                    }
                    currentDiscStart += currentDiscSize + 1;
                    offset += 1;
                }

                holder.bindItem(items.get(position - offset));
            } else {
                holder.bindItem(items.get(position));
            }
        } else if (holder instanceof BrowseItemHolderAlbumDiscHeader) {

            // Assign disc number
            BrowseItemHolderAlbumDiscHeader header = (BrowseItemHolderAlbumDiscHeader) holder;

            int currentDiscStart = 0;
            for (int discIndex = 0; discIndex < numDiscHeaders; discIndex++) {
                int currentDiscSize = discSizes.valueAt(discIndex);
                if (position == currentDiscStart) {
                    header.setDiscNumber(discSizes.keyAt(discIndex));
                    break;
                }
                currentDiscStart += currentDiscSize + 1;
            }
        }
    }

    enum AlbumViewType {
        DISC_HEADER,
        TRACK,
    }
}

File: app/src/main/java/agersant/polaris/features/browse/BrowseAdapterAlbum.kt
package agersant.polaris.features.browse

import agersant.polaris.CollectionItem
import agersant.polaris.PlaybackQueue
import agersant.polaris.api.API
import agersant.polaris.databinding.ViewBrowseAlbumDiscHeaderBinding
import agersant.polaris.databinding.ViewBrowseAlbumItemBinding
import agersant.polaris.databinding.ViewQueueStatusBinding
import android.util.SparseIntArray
import android.view.LayoutInflater
import android.view.ViewGroup

internal class BrowseAdapterAlbum(
    private val api: API,
    private val playbackQueue: PlaybackQueue,
) : BrowseAdapter() {
    // Key is disc number, value is number of tracks
    private val discSizes = SparseIntArray()
    private var numDiscHeaders = 0

    override fun updateItems(items: List<CollectionItem>) {
        for (item in items) {
            val discNumber = item.discNumber
            discSizes.put(discNumber, discSizes[discNumber, 0] + 1)
        }
        numDiscHeaders = discSizes.size()
        if (numDiscHeaders == 1) {
            numDiscHeaders = 0
        }

        this.items = items
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + numDiscHeaders
    }

    override fun getItemViewType(position: Int): Int {
        var currentDiscStart = 0
        for (discIndex in 0 until numDiscHeaders) {
            val currentDiscSize = discSizes.valueAt(discIndex)

            if (position == currentDiscStart) {
                return AlbumViewType.DISC_HEADER.ordinal
            } else if (position < currentDiscStart + currentDiscSize + 1) {
                return AlbumViewType.TRACK.ordinal
            }
            currentDiscStart += currentDiscSize + 1
        }

        return AlbumViewType.TRACK.ordinal
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BrowseItemHolder {
        val inflater = LayoutInflater.from(parent.context)
        val queueStatusBinding = ViewQueueStatusBinding.inflate(inflater, parent, false)

        return when (viewType) {
            AlbumViewType.DISC_HEADER.ordinal -> {
                val itemBinding = ViewBrowseAlbumDiscHeaderBinding.inflate(inflater, parent, false)
                BrowseItemHolderAlbumDiscHeader(api, playbackQueue, this, itemBinding, queueStatusBinding)
            }
            else -> {
                val itemBinding = ViewBrowseAlbumItemBinding.inflate(inflater, parent, false)
                BrowseItemHolderAlbumTrack(api, playbackQueue, this, itemBinding, queueStatusBinding)
            }
        }
    }

    override fun onBindViewHolder(holder: BrowseItemHolder, position: Int) {
        when (holder) {
            is BrowseItemHolderAlbumTrack -> {

                // Assign track item
                if (numDiscHeaders > 0) {
                    var offset = 1
                    var currentDiscStart = 0
                    for (discIndex in 0 until numDiscHeaders) {
                        val currentDiscSize = discSizes.valueAt(discIndex)
                        if (position < currentDiscStart + currentDiscSize + 1) {
                            break
                        }
                        currentDiscStart += currentDiscSize + 1
                        offset += 1
                    }
                    holder.bindItem(items[position - offset])
                } else {
                    holder.bindItem(items[position])
                }
            }
            is BrowseItemHolderAlbumDiscHeader -> {

                // Assign disc number
                var currentDiscStart = 0
                for (discIndex in 0 until numDiscHeaders) {
                    val currentDiscSize = discSizes.valueAt(discIndex)
                    if (position == currentDiscStart) {
                        holder.setDiscNumber(discSizes.keyAt(discIndex))
                        break
                    }
                    currentDiscStart += currentDiscSize + 1
                }
            }
        }
    }

    internal enum class AlbumViewType {
        DISC_HEADER,
        TRACK,
    }
}
